File: website/src/core/utils/datetime-utils.ts
/**
 * DateTime utility for consistent timezone handling across the website
 *
 * 핵심 원칙:
 * - DB에는 항상 UTC로 저장
 * - 화면에는 항상 로컬 시간으로 표시
 * - 날짜만 필요한 경우 타임존 변환 안 함
 *
 * 사용 예시:
 * ```typescript
 * // 저장
 * await supabase.insert({
 *   created_at: DateTimeUtils.toUtc(new Date())
 * });
 *
 * // 읽기
 * const data = await supabase.select();
 * const createdAt = DateTimeUtils.toLocal(data.created_at);
 * ```
 */
export class DateTimeUtils {
  /**
   * Converts Date to UTC ISO8601 string for database storage
   *
   * @param date - Local Date object
   * @returns UTC ISO8601 string (e.g., "2025-01-15T05:30:00.000Z")
   *
   * @example
   * const now = new Date(); // 2025-01-15 14:30:00 (KST +9)
   * DateTimeUtils.toUtc(now); // "2025-01-15T05:30:00.000Z" (UTC)
   */
  static toUtc(date: Date): string {
    return date.toISOString();
  }

  /**
   * Returns current time as UTC ISO8601 string
   *
   * @returns Current UTC timestamp
   *
   * @example
   * await supabase.insert({
   *   created_at: DateTimeUtils.nowUtc()
   * });
   */
  static nowUtc(): string {
    return new Date().toISOString();
  }

  /**
   * Converts UTC string from database to local Date object
   *
   * Handles both:
   * - ISO 8601 with timezone: "2025-01-15T05:30:00.000Z"
   * - Timestamp without timezone: "2025-10-27 17:54:41.715" (treats as UTC)
   *
   * @param utcString - UTC timestamp string
   * @returns Local Date object
   *
   * @example
   * const utcString1 = "2025-01-15T05:30:00.000Z"; // ISO 8601 UTC
   * const utcString2 = "2025-10-27 17:54:41.715"; // timestamp without timezone
   * const local1 = DateTimeUtils.toLocal(utcString1);
   * const local2 = DateTimeUtils.toLocal(utcString2);
   * // Both converted to local time correctly
   */
  static toLocal(utcString: string): Date {
    // If string doesn't contain timezone info (no Z, +, or -), add Z to force UTC parsing
    if (
      !utcString.includes('Z') &&
      !utcString.includes('+') &&
      !utcString.includes('-', utcString.length - 6)
    ) {
      // Replace space with T for ISO 8601 format, then add Z
      const isoFormat = utcString.replace(' ', 'T');
      return new Date(`${isoFormat}Z`);
    }
    return new Date(utcString);
  }

  /**
   * Safely converts UTC string to local Date object (null-safe version)
   *
   * Handles both:
   * - ISO 8601 with timezone: "2025-01-15T05:30:00.000Z"
   * - Timestamp without timezone: "2025-10-27 17:54:41.715" (treats as UTC)
   *
   * Returns null if:
   * - Input is null or undefined
   * - Input is empty string
   * - Parsing fails
   *
   * @param utcString - UTC timestamp string (nullable)
   * @returns Local Date object or null
   *
   * @example
   * const date = DateTimeUtils.toLocalSafe(json.updated_at);
   * if (date) {
   *   console.log('Updated:', DateTimeUtils.format(date));
   * }
   */
  static toLocalSafe(utcString?: string | null): Date | null {
    if (!utcString || utcString.trim() === '') return null;
    try {
      return this.toLocal(utcString);
    } catch (error) {
      console.error('Failed to parse date:', utcString, error);
      return null;
    }
  }

  /**
   * Converts Date to date-only string (yyyy-MM-dd)
   *
   * 중요: 타임존 변환을 하지 않습니다!
   * 날짜만 필요한 경우 사용하세요 (예: 생일, 계약일)
   *
   * @param date - Date object
   * @returns Date string in yyyy-MM-dd format
   *
   * @example
   * const birthday = new Date(1990, 4, 15); // May 15, 1990
   * DateTimeUtils.toDateOnly(birthday); // "1990-05-15"
   *
   * // ❌ 이렇게 하면 날짜가 바뀔 수 있음:
   * birthday.toISOString().split('T')[0];
   *
   * // ✅ 이렇게 해야 함:
   * DateTimeUtils.toDateOnly(birthday);
   */
  static toDateOnly(date: Date): string {
    const year = date.getFullYear();
    const month = String(date.getMonth() + 1).padStart(2, '0');
    const day = String(date.getDate()).padStart(2, '0');
    return `${year}-${month}-${day}`;
  }

  /**
   * Converts Date to Supabase RPC format (yyyy-MM-dd HH:mm:ss in UTC)
   *
   * Supabase의 일부 RPC 함수는 특정 포맷을 요구합니다.
   * 이 함수는 UTC로 변환 후 RPC 포맷으로 변환합니다.
   *
   * @param date - Local Date object
   * @returns UTC timestamp in Supabase RPC format
   *
   * @example
   * const now = new Date(); // 2025-01-15 14:30:00 (KST +9)
   * DateTimeUtils.toRpcFormat(now); // "2025-01-15 05:30:00" (UTC)
   *
   * await supabase.rpc('insert_cashier_amount_lines', {
   *   p_created_at: DateTimeUtils.toRpcFormat(new Date())
   * });
   */
  static toRpcFormat(date: Date): string {
    const utcString = date.toISOString();
    // Convert "2025-01-15T05:30:00.000Z" to "2025-01-15 05:30:00"
    return utcString.replace('T', ' ').split('.')[0];
  }

  /**
   * Formats Date for display on screen (yyyy-MM-dd HH:mm)
   *
   * 로컬 시간으로 표시됩니다.
   *
   * @param date - Local Date object
   * @returns Formatted string
   *
   * @example
   * const utcString = "2025-01-15T05:30:00.000Z"; // from DB
   * const local = DateTimeUtils.toLocal(utcString);
   * console.log(DateTimeUtils.format(local));
   * // Korea: "2025-01-15 14:30"
   * // Vietnam: "2025-01-15 12:30"
   */
  static format(date: Date): string {
    const year = date.getFullYear();
    const month = String(date.getMonth() + 1).padStart(2, '0');
    const day = String(date.getDate()).padStart(2, '0');
    const hours = String(date.getHours()).padStart(2, '0');
    const minutes = String(date.getMinutes()).padStart(2, '0');
    return `${year}-${month}-${day} ${hours}:${minutes}`;
  }

  /**
   * Formats Date with custom Intl.DateTimeFormat options
   *
   * @param date - Date object
   * @param options - Intl.DateTimeFormat options
   * @param locale - Locale string (default: 'ko-KR')
   * @returns Formatted string
   *
   * @example
   * const now = new Date();
   * DateTimeUtils.formatCustom(now, { dateStyle: 'long' }); // "2025년 1월 15일"
   * DateTimeUtils.formatCustom(now, { timeStyle: 'short' }); // "오후 2:30"
   * DateTimeUtils.formatCustom(now, { dateStyle: 'medium', timeStyle: 'short' }, 'en-US');
   * // "Jan 15, 2025, 2:30 PM"
   */
  static formatCustom(
    date: Date,
    options: Intl.DateTimeFormatOptions = {},
    locale: string = 'ko-KR'
  ): string {
    return new Intl.DateTimeFormat(locale, options).format(date);
  }

  /**
   * Formats date-only display (yyyy-MM-dd)
   *
   * @param date - Date object
   * @returns Date string
   *
   * @example
   * const date = new Date();
   * DateTimeUtils.formatDateOnly(date); // "2025-01-15"
   */
  static formatDateOnly(date: Date): string {
    return this.toDateOnly(date);
  }

  /**
   * Formats time-only display (HH:mm)
   *
   * @param date - Date object
   * @returns Time string
   *
   * @example
   * const time = new Date();
   * DateTimeUtils.formatTimeOnly(time); // "14:30"
   */
  static formatTimeOnly(date: Date): string {
    const hours = String(date.getHours()).padStart(2, '0');
    const minutes = String(date.getMinutes()).padStart(2, '0');
    return `${hours}:${minutes}`;
  }

  /**
   * Formats date with relative time (e.g., "2 hours ago", "yesterday")
   *
   * @param date - Date object
   * @param locale - Locale string (default: 'ko-KR')
   * @returns Relative time string
   *
   * @example
   * const now = new Date();
   * const twoHoursAgo = new Date(now.getTime() - 2 * 60 * 60 * 1000);
   * DateTimeUtils.formatRelative(twoHoursAgo); // "2시간 전"
   * DateTimeUtils.formatRelative(twoHoursAgo, 'en-US'); // "2 hours ago"
   */
  static formatRelative(date: Date, locale: string = 'ko-KR'): string {
    const now = new Date();
    const diffMs = now.getTime() - date.getTime();
    const diffSec = Math.floor(diffMs / 1000);
    const diffMin = Math.floor(diffSec / 60);
    const diffHour = Math.floor(diffMin / 60);
    const diffDay = Math.floor(diffHour / 24);

    const rtf = new Intl.RelativeTimeFormat(locale, { numeric: 'auto' });

    if (diffSec < 60) {
      return rtf.format(-diffSec, 'second');
    } else if (diffMin < 60) {
      return rtf.format(-diffMin, 'minute');
    } else if (diffHour < 24) {
      return rtf.format(-diffHour, 'hour');
    } else if (diffDay < 30) {
      return rtf.format(-diffDay, 'day');
    } else {
      return this.format(date);
    }
  }

  /**
   * Checks if two dates are on the same day (ignores time)
   *
   * @param date1 - First date
   * @param date2 - Second date
   * @returns True if same day
   *
   * @example
   * const date1 = new Date('2025-01-15 10:00:00');
   * const date2 = new Date('2025-01-15 20:00:00');
   * DateTimeUtils.isSameDay(date1, date2); // true
   */
  static isSameDay(date1: Date, date2: Date): boolean {
    return (
      date1.getFullYear() === date2.getFullYear() &&
      date1.getMonth() === date2.getMonth() &&
      date1.getDate() === date2.getDate()
    );
  }

  /**
   * Adds days to a date
   *
   * @param date - Base date
   * @param days - Number of days to add (can be negative)
   * @returns New date
   *
   * @example
   * const today = new Date('2025-01-15');
   * const tomorrow = DateTimeUtils.addDays(today, 1); // 2025-01-16
   * const yesterday = DateTimeUtils.addDays(today, -1); // 2025-01-14
   */
  static addDays(date: Date, days: number): Date {
    const result = new Date(date);
    result.setDate(result.getDate() + days);
    return result;
  }

  /**
   * Gets start of day (00:00:00.000)
   *
   * @param date - Date object
   * @returns Date at start of day
   *
   * @example
   * const now = new Date('2025-01-15 14:30:00');
   * DateTimeUtils.startOfDay(now); // 2025-01-15 00:00:00
   */
  static startOfDay(date: Date): Date {
    const result = new Date(date);
    result.setHours(0, 0, 0, 0);
    return result;
  }

  /**
   * Gets end of day (23:59:59.999)
   *
   * @param date - Date object
   * @returns Date at end of day
   *
   * @example
   * const now = new Date('2025-01-15 14:30:00');
   * DateTimeUtils.endOfDay(now); // 2025-01-15 23:59:59.999
   */
  static endOfDay(date: Date): Date {
    const result = new Date(date);
    result.setHours(23, 59, 59, 999);
    return result;
  }

  /**
   * Formats date for dashboard transactions (Nov 1, 12:00 AM)
   *
   * @param dateString - Date string (ISO or any format parseable by Date)
   * @returns Formatted string (e.g., "Nov 1, 12:00 AM")
   *
   * @example
   * DateTimeUtils.formatTransactionDate("2025-01-15T14:30:00.000Z");
   * // "Jan 15, 2:30 PM"
   */
  static formatTransactionDate(dateString: string): string {
    try {
      const date = new Date(dateString);

      // Check if date is valid
      if (isNaN(date.getTime())) {
        return 'Invalid Date';
      }

      // Format: Nov 1, 12:00 AM
      const options: Intl.DateTimeFormatOptions = {
        month: 'short',
        day: 'numeric',
        hour: 'numeric',
        minute: '2-digit',
        hour12: true,
      };

      return date.toLocaleString('en-US', options);
    } catch (error) {
      console.error('Date formatting error:', error);
      return 'Invalid Date';
    }
  }
}

File: myFinance_improved_V2/lib/core/utils/datetime_utils.dart
import 'package:intl/intl.dart';

/// DateTime utility for consistent timezone handling across the app
///
/// **핵심 원칙:**
/// - DB에는 항상 UTC로 저장
/// - 화면에는 항상 로컬 시간으로 표시
/// - 날짜만 필요한 경우 타임존 변환 안 함
///
/// **사용 예시:**
/// ```dart
/// // 저장
/// await supabase.insert({
///   'created_at': DateTimeUtils.toUtc(DateTime.now())
/// });
///
/// // 읽기
/// final data = await supabase.select();
/// final createdAt = DateTimeUtils.toLocal(data['created_at']);
/// ```
class DateTimeUtils {

  /// Converts DateTime to UTC ISO8601 string for database storage
  ///
  /// Example:
  /// ```dart
  /// final now = DateTime.now(); // 2025-01-15 14:30:00 (KST +9)
  /// DateTimeUtils.toUtc(now);   // "2025-01-15T05:30:00.000Z" (UTC)
  /// ```
  static String toUtc(DateTime dateTime) {
    return dateTime.toUtc().toIso8601String();
  }

  /// Returns current time as UTC ISO8601 string
  ///
  /// Example:
  /// ```dart
  /// await supabase.insert({
  ///   'created_at': DateTimeUtils.nowUtc()
  /// });
  /// ```
  static String nowUtc() {
    return DateTime.now().toUtc().toIso8601String();
  }

  /// Converts UTC string from database to local DateTime
  ///
  /// Handles both:
  /// - ISO 8601 with timezone: "2025-01-15T05:30:00.000Z"
  /// - Timestamp without timezone: "2025-10-27 17:54:41.715" (treats as UTC)
  ///
  /// Example:
  /// ```dart
  /// final utcString1 = "2025-01-15T05:30:00.000Z"; // ISO 8601 UTC
  /// final utcString2 = "2025-10-27 17:54:41.715"; // timestamp without timezone
  /// final local1 = DateTimeUtils.toLocal(utcString1);
  /// final local2 = DateTimeUtils.toLocal(utcString2);
  /// // Both converted to local time correctly
  /// ```
  ///
  /// Throws FormatException if string is invalid
  static DateTime toLocal(String utcString) {
    // Handle empty string
    if (utcString.isEmpty) {
      throw const FormatException('Cannot parse empty string as DateTime');
    }

    try {
      // If string doesn't contain timezone info (no Z, +, or -), add Z to force UTC parsing
      if (!utcString.contains('Z') &&
          !utcString.contains('+') &&
          !utcString.contains('-', utcString.length - 6)) {
        // Replace space with T for ISO 8601 format, then add Z
        final isoFormat = utcString.replaceFirst(' ', 'T');
        final result = DateTime.parse('${isoFormat}Z').toLocal();
        return result;
      }
      final result = DateTime.parse(utcString).toLocal();
      return result;
    } catch (e) {
      // Fallback: Remove timezone offset and treat as UTC
      try {
        // Remove any timezone offset (e.g., +00:00, -06:00, etc.)
        final cleanString = utcString
            .replaceAll(RegExp(r'[+-]\d{2}:\d{2}$'), '')
            .replaceAll('Z', '')
            .replaceFirst(' ', 'T');

        if (cleanString.isEmpty) {
          throw const FormatException('String became empty after cleaning');
        }

        final result = DateTime.parse('${cleanString}Z').toLocal();
        return result;
      } catch (fallbackError) {
        rethrow;
      }
    }
  }

  /// Safely converts UTC string to local DateTime (null-safe version)
  ///
  /// Handles both:
  /// - ISO 8601 with timezone: "2025-01-15T05:30:00.000Z"
  /// - Timestamp without timezone: "2025-10-27 17:54:41.715" (treats as UTC)
  ///
  /// Returns null if:
  /// - Input is null
  /// - Input is empty string
  /// - Parsing fails
  ///
  /// Example:
  /// ```dart
  /// final date = DateTimeUtils.toLocalSafe(json['updated_at']);
  /// if (date != null) {
  ///   print('Updated: ${DateTimeUtils.format(date)}');
  /// }
  /// ```
  static DateTime? toLocalSafe(String? utcString) {
    if (utcString == null || utcString.isEmpty) return null;
    try {
      return toLocal(utcString);
    } catch (e) {
      return null;
    }
  }

  /// Converts DateTime to date-only string (yyyy-MM-dd)
  ///
  /// **중요:** 타임존 변환을 하지 않습니다!
  /// 날짜만 필요한 경우 사용하세요 (예: 생일, 계약일)
  ///
  /// Example:
  /// ```dart
  /// final birthday = DateTime(1990, 5, 15);
  /// DateTimeUtils.toDateOnly(birthday); // "1990-05-15"
  ///
  /// // ❌ 이렇게 하면 날짜가 바뀔 수 있음:
  /// birthday.toUtc().toIso8601String().split('T')[0];
  ///
  /// // ✅ 이렇게 해야 함:
  /// DateTimeUtils.toDateOnly(birthday);
  /// ```
  static String toDateOnly(DateTime date) {
    return '${date.year}-${date.month.toString().padLeft(2, '0')}-${date.day.toString().padLeft(2, '0')}';
  }

  /// Converts DateTime to Supabase RPC format (yyyy-MM-dd HH:mm:ss in UTC)
  ///
  /// Supabase의 일부 RPC 함수는 특정 포맷을 요구합니다.
  /// 이 함수는 UTC로 변환 후 RPC 포맷으로 변환합니다.
  ///
  /// Example:
  /// ```dart
  /// final now = DateTime.now(); // 2025-01-15 14:30:00 (KST +9)
  /// DateTimeUtils.toRpcFormat(now); // "2025-01-15 05:30:00" (UTC)
  ///
  /// await supabase.rpc('insert_cashier_amount_lines', {
  ///   'p_created_at': DateTimeUtils.toRpcFormat(DateTime.now())
  /// });
  /// ```
  static String toRpcFormat(DateTime dateTime) {
    final utc = dateTime.toUtc();
    return utc.toIso8601String().replaceFirst('T', ' ').split('.')[0];
  }

  /// Formats DateTime for display on screen (yyyy-MM-dd HH:mm)
  ///
  /// **로컬 시간**으로 표시됩니다.
  ///
  /// Example:
  /// ```dart
  /// final utcString = "2025-01-15T05:30:00.000Z"; // from DB
  /// final local = DateTimeUtils.toLocal(utcString);
  /// print(DateTimeUtils.format(local));
  /// // Korea: "2025-01-15 14:30"
  /// // Vietnam: "2025-01-15 12:30"
  /// ```
  static String format(DateTime dateTime) {
    return '${dateTime.year}-${dateTime.month.toString().padLeft(2, '0')}-${dateTime.day.toString().padLeft(2, '0')} ${dateTime.hour.toString().padLeft(2, '0')}:${dateTime.minute.toString().padLeft(2, '0')}';
  }

  /// Formats DateTime with custom pattern using intl package
  ///
  /// Example:
  /// ```dart
  /// final now = DateTime.now();
  /// DateTimeUtils.formatCustom(now, 'yyyy년 MM월 dd일'); // "2025년 01월 15일"
  /// DateTimeUtils.formatCustom(now, 'MMM dd, yyyy'); // "Jan 15, 2025"
  /// DateTimeUtils.formatCustom(now, 'HH:mm:ss'); // "14:30:00"
  /// ```
  static String formatCustom(DateTime dateTime, String pattern) {
    final formatter = DateFormat(pattern);
    return formatter.format(dateTime);
  }

  /// Formats date-only display (yyyy-MM-dd)
  ///
  /// Example:
  /// ```dart
  /// final date = DateTime.now();
  /// DateTimeUtils.formatDateOnly(date); // "2025-01-15"
  /// ```
  static String formatDateOnly(DateTime dateTime) {
    return toDateOnly(dateTime);
  }

  /// Formats time-only display (HH:mm)
  ///
  /// Example:
  /// ```dart
  /// final time = DateTime.now();
  /// DateTimeUtils.formatTimeOnly(time); // "14:30"
  /// ```
  static String formatTimeOnly(DateTime dateTime) {
    return '${dateTime.hour.toString().padLeft(2, '0')}:${dateTime.minute.toString().padLeft(2, '0')}';
  }

  /// Converts DateTime to ISO 8601 string with local timezone offset
  ///
  /// This format includes the timezone offset to preserve exact moment in time
  /// while maintaining the local time representation.
  ///
  /// Format: "yyyy-MM-ddTHH:mm:ss+HH:mm" (e.g., "2024-11-15T10:30:25+07:00")
  ///
  /// Example:
  /// ```dart
  /// final now = DateTime.now(); // 2025-01-15 14:30:25 (KST +9)
  /// DateTimeUtils.toLocalWithOffset(now); // "2025-01-15T14:30:25+09:00"
  ///
  /// // Vietnam time
  /// final vietnamTime = DateTime.now(); // 2024-11-15 10:30:25 (ICT +7)
  /// DateTimeUtils.toLocalWithOffset(vietnamTime); // "2024-11-15T10:30:25+07:00"
  ///
  /// // For RPC calls
  /// await supabase.rpc('user_shift_overview_v3', {
  ///   'p_request_time': DateTimeUtils.toLocalWithOffset(DateTime.now())
  /// });
  /// ```
  static String toLocalWithOffset(DateTime dateTime) {
    // Get timezone offset
    final offset = dateTime.timeZoneOffset;
    final hours = offset.inHours;
    final minutes = offset.inMinutes.remainder(60);

    // Format offset as +HH:mm or -HH:mm
    final sign = hours >= 0 ? '+' : '-';
    final absHours = hours.abs().toString().padLeft(2, '0');
    final absMinutes = minutes.abs().toString().padLeft(2, '0');
    final offsetString = '$sign$absHours:$absMinutes';

    // Format datetime as ISO 8601 with offset
    final year = dateTime.year.toString();
    final month = dateTime.month.toString().padLeft(2, '0');
    final day = dateTime.day.toString().padLeft(2, '0');
    final hour = dateTime.hour.toString().padLeft(2, '0');
    final minute = dateTime.minute.toString().padLeft(2, '0');
    final second = dateTime.second.toString().padLeft(2, '0');

    return '$year-$month-${day}T$hour:$minute:$second$offsetString';
  }

  /// Formats current local DateTime as 'yyyy-MM-dd HH:mm:ss' string
  ///
  /// This format is commonly required by Supabase RPC functions
  /// that need local time with timezone parameter.
  ///
  /// Example:
  /// ```dart
  /// final localTime = DateTimeUtils.formatLocalTimestamp();
  /// // "2025-01-15 14:30:25"
  ///
  /// await supabase.rpc('create_store_shift', {
  ///   'p_time': localTime,
  ///   'p_timezone': DateTimeUtils.getLocalTimezone(),
  /// });
  /// ```
  static String formatLocalTimestamp([DateTime? dateTime]) {
    final now = dateTime ?? DateTime.now();
    return '${now.year}-${now.month.toString().padLeft(2, '0')}-${now.day.toString().padLeft(2, '0')} '
        '${now.hour.toString().padLeft(2, '0')}:${now.minute.toString().padLeft(2, '0')}:${now.second.toString().padLeft(2, '0')}';
  }

  /// Gets the user's local timezone in IANA format (e.g., "Asia/Seoul", "Asia/Ho_Chi_Minh")
  ///
  /// This timezone string is used for RPC functions that need to handle timezone-aware calculations.
  /// Since Dart doesn't provide direct access to IANA timezone names, we derive it from the offset.
  ///
  /// **Note**: This is a best-effort approximation. Multiple timezones can share the same offset.
  /// For production, consider using platform-specific APIs or a timezone package.
  ///
  /// Example:
  /// ```dart
  /// final timezone = DateTimeUtils.getLocalTimezone();
  /// // Korea (UTC+9): "Asia/Seoul"
  /// // Vietnam (UTC+7): "Asia/Ho_Chi_Minh"
  /// // UTC: "UTC"
  ///
  /// await supabase.rpc('user_shift_overview_v3', {
  ///   'p_timezone': timezone,
  /// });
  /// ```
  static String getLocalTimezone() {
    final now = DateTime.now();
    final offset = now.timeZoneOffset;

    // Get timezone name from DateTime
    final timeZoneName = now.timeZoneName;

    // If timezone name is available and not just an offset, use it
    if (timeZoneName.isNotEmpty && !timeZoneName.contains('+') && !timeZoneName.contains('-')) {
      // Common timezone abbreviations to IANA mapping
      const timezoneMap = {
        'KST': 'Asia/Seoul',
        'JST': 'Asia/Tokyo',
        'ICT': 'Asia/Ho_Chi_Minh',
        'CST': 'Asia/Shanghai',
        'PST': 'America/Los_Angeles',
        'EST': 'America/New_York',
        'GMT': 'Europe/London',
        'UTC': 'UTC',
      };

      if (timezoneMap.containsKey(timeZoneName)) {
        return timezoneMap[timeZoneName]!;
      }
    }

    // Fallback: Map offset to most common timezone
    // This is not perfect but works for most cases
    final hours = offset.inHours;
    final minutes = offset.inMinutes.remainder(60);

    // Format: +HH:MM or -HH:MM
    final sign = hours >= 0 ? '+' : '-';
    final absHours = hours.abs().toString().padLeft(2, '0');
    final absMinutes = minutes.abs().toString().padLeft(2, '0');

    // Map common offsets to IANA timezones
    const offsetToTimezone = {
      '+09:00': 'Asia/Seoul',      // Korea, Japan
      '+08:00': 'Asia/Shanghai',   // China, Singapore
      '+07:00': 'Asia/Ho_Chi_Minh', // Vietnam, Thailand
      '+05:30': 'Asia/Kolkata',    // India
      '+00:00': 'UTC',             // UTC
      '-05:00': 'America/New_York', // US Eastern
      '-08:00': 'America/Los_Angeles', // US Pacific
    };

    final offsetString = '$sign$absHours:$absMinutes';
    return offsetToTimezone[offsetString] ?? 'UTC';
  }
}
